File: packages/graph-grammar/src/engine.ts
import type { Grammar, Rule, Match, Graph, RewriteResult } from './types.ts'
import { GraphIndex, cloneGraph } from './graph.ts'
import { findMatches, findOneMatch, hasMatch, countMatches } from './match.ts'
import { applyRule, type RewriteContext } from './rewrite.ts'
import { RNG } from './util.ts'

// ============================================================================
// The grammar engine. Drives application of a rule set to a host graph under a
// chosen strategy, with reproducible stochastic behaviour.
//
// Performance contract: a single rewrite step costs O(match-find), independent
// of the total number of matches in the graph. We never materialise the whole
// graph per step and we never enumerate the full match set just to apply one
// (single-match strategies use a randomised first-match search instead).
// ============================================================================

const EMPTY_RESULT: () => RewriteResult = () => ({
  applied: false,
  createdNodes: [],
  createdEdges: [],
  deletedNodes: [],
  deletedEdges: [],
})

export class Engine {
  index: GraphIndex
  grammar: Grammar
  ctx: RewriteContext
  steps = 0
  private appCount = new Map<string, number>()
  private seqPtr = 0

  constructor (grammar: Grammar, start?: Graph) {
    this.grammar = grammar
    // Clone so the engine never mutates the grammar's start/axiom in place ,
    // rewrites relabel and delete nodes, and the start must survive a reset.
    this.index = new GraphIndex(cloneGraph(start ?? grammar.start))
    this.ctx = { rng: new RNG(grammar.config.seed), counter: { value: 0 } }
  }

  reset (start?: Graph) {
    this.index = new GraphIndex(cloneGraph(start ?? this.grammar.start))
    this.ctx = { rng: new RNG(this.grammar.config.seed), counter: { value: 0 } }
    this.steps = 0
    this.appCount.clear()
    this.seqPtr = 0
  }

  get graph (): Graph {
    return this.index.toGraph()
  }

  private enabledRules (): Rule[] {
    return this.grammar.rules.filter(
      (r) =>
        r.enabled &&
        r.lhs.nodes.length > 0 &&
        (r.maxApplications === 0 || (this.appCount.get(r.id) ?? 0) < r.maxApplications) &&
        this.fitsNodeBudget(r)
    )
  }

  /** Net change in node count if this rule fires once: created − deleted. */
  nodeDelta (rule: Rule): number {
    const preserved = new Set(rule.rhs.nodes.map((n) => n.mapFrom).filter(Boolean) as string[])
    const created = rule.rhs.nodes.filter((n) => !n.mapFrom).length
    const deleted = rule.lhs.nodes.filter((n) => !preserved.has(n.id)).length
    return created - deleted
  }

  /** Would applying this rule keep the graph within maxNodes? Net-zero and
   *  shrinking rules always fit, so generation can still resolve at the cap. */
  fitsNodeBudget (rule: Rule): boolean {
    const cap = this.grammar.config.maxNodes
    return cap <= 0 || this.index.nodes.size + this.nodeDelta(rule) <= cap
  }

  /** True if the rule would grow past maxNodes (used by the UI to show why a
   *  matching rule can't currently fire). */
  isBlockedByNodeCap (rule: Rule): boolean {
    const cap = this.grammar.config.maxNodes
    return cap > 0 && this.index.nodes.size + this.nodeDelta(rule) > cap
  }

  /** Global-form NAC: block the rule if any NAC pattern exists in the host. */
  private nacBlocked (rule: Rule): boolean {
    if (!rule.nac || rule.nac.length === 0) return false
    for (const nac of rule.nac) if (nac.nodes.length && hasMatch(nac, this.index)) return true
    return false
  }

  /** A single random match for a rule (NAC-aware), or null. */
  private oneMatch (rule: Rule): Match | null {
    if (this.nacBlocked(rule)) return null
    return findOneMatch(rule.id, rule.lhs, this.index, this.ctx.rng)
  }

  step (): RewriteResult {
    const cfg = this.grammar.config
    // No blanket node-cap gate here: rules that would grow past maxNodes are
    // excluded by enabledRules()/fitsNodeBudget(), while net-zero / shrinking
    // rules still fire so generation can resolve at the cap.

    if (cfg.strategy === 'maximal') {
      const rule = this.pickMaximalRule()
      return rule ? this.applyMaximal(rule) : EMPTY_RESULT()
    }

    // Find a (rule, match) pair in one pass , no separate applicability probe.
    const picked = this.pickRuleAndMatch()
    if (!picked) return EMPTY_RESULT()
    const { rule, match: m } = picked

    // probability gate: counts as a (no-op) step, lets the run loop try again
    if (rule.probability < 1 && this.ctx.rng.next() > rule.probability) {
      const r = EMPTY_RESULT()
      r.ruleId = rule.id
      return r
    }

    const changes = applyRule(this.index, rule, m, this.ctx)
    this.appCount.set(rule.id, (this.appCount.get(rule.id) ?? 0) + 1)
    this.steps++
    return { applied: true, ruleId: rule.id, match: m, ...changes }
  }

  /**
   * Select a rule AND one of its matches together, so we pay for at most one
   * match search per rule tried (and usually exactly one for the whole step).
   */
  private pickRuleAndMatch (): { rule: Rule; match: Match } | null {
    const rules = this.enabledRules()
    if (rules.length === 0) return null
    const cfg = this.grammar.config

    if (cfg.strategy === 'priority') {
      const sorted = [...rules].sort((a, b) => b.priority - a.priority)
      for (const r of sorted) {
        const m = this.oneMatch(r)
        if (m) return { rule: r, match: m }
      }
      return null
    }
    if (cfg.strategy === 'sequential') {
      for (let i = 0; i < rules.length; i++) {
        const r = rules[(this.seqPtr + i) % rules.length]
        const m = this.oneMatch(r)
        if (m) {
          this.seqPtr = (this.seqPtr + i + 1) % rules.length
          return { rule: r, match: m }
        }
      }
      return null
    }
    // random: draw rules in weight-biased order; first one that matches wins.
    // This is a weighted random choice among the *applicable* rules without
    // probing every rule up front.
    const pool = rules.slice()
    const weights = pool.map((r) => r.weight)
    while (pool.length) {
      const idx = this.ctx.rng.weightedIndex(weights)
      const r = pool[idx]
      const m = this.oneMatch(r)
      if (m) return { rule: r, match: m }
      pool.splice(idx, 1)
      weights.splice(idx, 1)
    }
    return null
  }

  /** For the maximal strategy: weighted pick among rules that currently match. */
  private pickMaximalRule (): Rule | null {
    const rules = this.enabledRules().filter(
      (r) => !this.nacBlocked(r) && findMatches(r.id, r.lhs, this.index, { limit: 1 }).length > 0
    )
    if (!rules.length) return null
    const idx = this.ctx.rng.weightedIndex(rules.map((r) => r.weight))
    return rules[idx]
  }

  /** Apply as many non-overlapping matches of one rule as possible this step. */
  private applyMaximal (rule: Rule): RewriteResult {
    const matches = findMatches(rule.id, rule.lhs, this.index)
    if (!matches.length) return EMPTY_RESULT()
    this.ctx.rng.shuffle(matches)
    const used = new Set<string>()
    const agg = EMPTY_RESULT()
    let count = 0
    let firstMatch: Match | undefined
    for (const m of matches) {
      const nodes = Object.values(m.nodeMap)
      if (nodes.some((n) => used.has(n))) continue
      if (rule.probability < 1 && this.ctx.rng.next() > rule.probability) continue
      if (rule.maxApplications && (this.appCount.get(rule.id) ?? 0) >= rule.maxApplications) break
      if (!this.fitsNodeBudget(rule)) break // stop before exceeding maxNodes
      const changes = applyRule(this.index, rule, m, this.ctx)
      for (const n of nodes) used.add(n)
      agg.createdNodes.push(...changes.createdNodes)
      agg.createdEdges.push(...changes.createdEdges)
      agg.deletedNodes.push(...changes.deletedNodes)
      agg.deletedEdges.push(...changes.deletedEdges)
      this.appCount.set(rule.id, (this.appCount.get(rule.id) ?? 0) + 1)
      if (!firstMatch) firstMatch = m
      count++
    }
    if (count) this.steps++
    agg.applied = count > 0
    agg.ruleId = rule.id
    agg.match = firstMatch
    return agg
  }

  /**
   * Run repeatedly until no rule applies or a bound is hit. Returns the number
   * of applied steps. The optional callback observes progress (don't pass one
   * for max throughput , it forces work per step).
   */
  run (maxSteps?: number, onStep?: (r: RewriteResult, i: number) => void): number {
    const cfg = this.grammar.config
    const requested = maxSteps ?? cfg.maxSteps
    // maxSteps <= 0 means "no step cap". For this synchronous path we still keep
    // a hard safety bound so an unbounded grammar can't freeze the page , when a
    // node cap is set the loop is bounded by that instead.
    // Uncapped runs keep a hard safety bound so the synchronous loop can't hang;
    // maxNodes is enforced per-rule by fitsNodeBudget, so the run naturally stops
    // once only non-growing rules remain and they reach a fixpoint.
    const cap = requested > 0 ? requested : 500_000
    let applied = 0
    let emptyStreak = 0
    for (let i = 0; i < cap; i++) {
      const r = this.step()
      if (!r.applied) {
        // probability skips return a ruleId; allow a bounded number of retries
        if (r.ruleId && emptyStreak < 64) {
          emptyStreak++
          onStep?.(r, i)
          continue
        }
        break
      }
      emptyStreak = 0
      applied++
      onStep?.(r, i)
    }
    return applied
  }

  /**
   * Match counts for the UI badges, each capped so a huge graph can't stall the
   * render loop. Returns the count or `cap` (caller renders "cap+").
   */
  matchCounts (cap = 200): Record<string, number> {
    const out: Record<string, number> = {}
    for (const r of this.grammar.rules) {
      if (!r.enabled || r.lhs.nodes.length === 0 || this.nacBlocked(r)) {
        out[r.id] = 0
        continue
      }
      out[r.id] = countMatches(r.lhs, this.index, cap)
    }
    return out
  }
}

File: packages/graph-grammar-native/src/engine.rs
use std::collections::{HashMap, HashSet};

use serde::Serialize;

use crate::graph::Host;
use crate::matching::{find_matches, find_one_match};
use crate::rewrite::apply_rule_at;
use crate::rng::Rng;
use crate::schema::{parse_validated, schema_check};
use crate::types::{ApiError, Grammar, Graph, MatchResult, Rule};

// ============================================================================
// Stateful engine ,a port of engine.ts. Drives a rule set over the host graph
// under a strategy (random/priority/sequential/maximal), with reproducible RNG
// seeded from config.seed. Holds the graph, RNG, per-rule application counts and
// the sequential pointer. One step costs one match search.
// ============================================================================

#[derive(Serialize, Default)]
pub(crate) struct StepResult {
    applied: bool,
    #[serde(rename = "ruleId", skip_serializing_if = "Option::is_none")]
    rule_id: Option<String>,
    #[serde(rename = "createdNodes")]
    created_nodes: Vec<String>,
    #[serde(rename = "createdEdges")]
    created_edges: Vec<String>,
    #[serde(rename = "deletedNodes")]
    deleted_nodes: Vec<String>,
    #[serde(rename = "deletedEdges")]
    deleted_edges: Vec<String>,
}

/// Opaque stateful engine handle (exposed across the FFI as `Engine*`).
pub struct Engine {
    grammar: Grammar,
    pub(crate) host: Host,
    rng: Rng,
    counter: i64,
    id_seq: i64,
    steps: i64,
    app_count: HashMap<String, i64>,
    seq_ptr: usize,
}

impl Engine {
    fn new(grammar: Grammar, start: Option<Graph>) -> Self {
        let start_graph = start.unwrap_or_else(|| grammar.start.clone());
        let rng = Rng::new(grammar.config.seed as u32);
        Engine {
            host: Host::from_graph(start_graph),
            grammar,
            rng,
            counter: 0,
            id_seq: 0,
            steps: 0,
            app_count: HashMap::new(),
            seq_ptr: 0,
        }
    }

    fn node_delta(&self, rule: &Rule) -> i64 {
        let preserved: HashSet<&str> =
            rule.rhs.nodes.iter().filter_map(|n| n.map_from.as_deref()).collect();
        let created = rule.rhs.nodes.iter().filter(|n| n.map_from.is_none()).count() as i64;
        let deleted =
            rule.lhs.nodes.iter().filter(|n| !preserved.contains(n.id.as_str())).count() as i64;
        created - deleted
    }

    fn fits_node_budget(&self, rule: &Rule) -> bool {
        let cap = self.grammar.config.max_nodes;
        cap <= 0 || self.host.nodes.len() as i64 + self.node_delta(rule) <= cap
    }

    fn enabled_rules(&self) -> Vec<usize> {
        (0..self.grammar.rules.len())
            .filter(|&i| {
                let r = &self.grammar.rules[i];
                r.enabled
                    && !r.lhs.nodes.is_empty()
                    && (r.max_applications == 0
                        || self.app_count.get(&r.id).copied().unwrap_or(0) < r.max_applications)
                    && self.fits_node_budget(r)
            })
            .collect()
    }

    fn nac_blocked(&self, ri: usize) -> bool {
        self.grammar.rules[ri]
            .nac
            .iter()
            .any(|nac| !nac.nodes.is_empty() && find_one_match(nac, &self.host, None).is_some())
    }

    /// A single random match for a rule (NAC-aware), or None. Consumes RNG.
    fn one_match(&mut self, ri: usize) -> Option<MatchResult> {
        if self.nac_blocked(ri) {
            return None;
        }
        let lhs = &self.grammar.rules[ri].lhs;
        find_one_match(lhs, &self.host, Some(&mut self.rng))
    }

    fn pick_rule_and_match(&mut self) -> Option<(usize, MatchResult)> {
        let rules = self.enabled_rules();
        if rules.is_empty() {
            return None;
        }
        match self.grammar.config.strategy.as_str() {
            "priority" => {
                let mut sorted = rules;
                // stable sort by priority desc (mirrors [...].sort((a,b)=>b.priority-a.priority))
                sorted.sort_by(|&a, &b| {
                    self.grammar.rules[b]
                        .priority
                        .partial_cmp(&self.grammar.rules[a].priority)
                        .unwrap_or(std::cmp::Ordering::Equal)
                });
                for ri in sorted {
                    if let Some(m) = self.one_match(ri) {
                        return Some((ri, m));
                    }
                }
                None
            }
            "sequential" => {
                let n = rules.len();
                for i in 0..n {
                    let ri = rules[(self.seq_ptr + i) % n];
                    if let Some(m) = self.one_match(ri) {
                        self.seq_ptr = (self.seq_ptr + i + 1) % n;
                        return Some((ri, m));
                    }
                }
                None
            }
            // random (default): weight-biased draw among applicable rules; first
            // that matches wins, without probing every rule up front.
            _ => {
                let mut pool = rules;
                let mut weights: Vec<f64> =
                    pool.iter().map(|&ri| self.grammar.rules[ri].weight).collect();
                while !pool.is_empty() {
                    let idx = self.rng.weighted_index(&weights);
                    let ri = pool[idx];
                    if let Some(m) = self.one_match(ri) {
                        return Some((ri, m));
                    }
                    pool.remove(idx);
                    weights.remove(idx);
                }
                None
            }
        }
    }

    fn pick_maximal_rule(&mut self) -> Option<usize> {
        // Filter is deterministic (no RNG); only the weighted pick draws.
        let candidates: Vec<usize> = self
            .enabled_rules()
            .into_iter()
            .filter(|&ri| {
                !self.nac_blocked(ri)
                    && !find_matches(&self.grammar.rules[ri].lhs, &self.host, 1, None).is_empty()
            })
            .collect();
        if candidates.is_empty() {
            return None;
        }
        let weights: Vec<f64> = candidates.iter().map(|&ri| self.grammar.rules[ri].weight).collect();
        let idx = self.rng.weighted_index(&weights);
        Some(candidates[idx])
    }

    fn apply_maximal(&mut self, ri: usize) -> Result<StepResult, ApiError> {
        // All matches, deterministic order, then shuffled (mirrors applyMaximal).
        let mut matches = find_matches(&self.grammar.rules[ri].lhs, &self.host, 0, None);
        if matches.is_empty() {
            return Ok(StepResult::default());
        }
        self.rng.shuffle(&mut matches);

        let mut used: HashSet<String> = HashSet::new();
        let mut agg = StepResult::default();
        let mut count = 0i64;
        for m in &matches {
            let nodes: Vec<String> = m.node_map.values().cloned().collect();
            if nodes.iter().any(|n| used.contains(n)) {
                continue;
            }
            let prob = self.grammar.rules[ri].probability;
            if prob < 1.0 && self.rng.next() > prob {
                continue;
            }
            let max_app = self.grammar.rules[ri].max_applications;
            let id = self.grammar.rules[ri].id.clone();
            if max_app != 0 && self.app_count.get(&id).copied().unwrap_or(0) >= max_app {
                break;
            }
            if !self.fits_node_budget(&self.grammar.rules[ri]) {
                break;
            }
            let out = apply_rule_at(
                &mut self.host,
                &self.grammar.rules[ri],
                m,
                Some(&mut self.rng),
                &mut self.counter,
                &mut self.id_seq,
            )?;
            for n in nodes {
                used.insert(n);
            }
            agg.created_nodes.extend(out.created_nodes);
            agg.created_edges.extend(out.created_edges);
            agg.deleted_nodes.extend(out.deleted_nodes);
            agg.deleted_edges.extend(out.deleted_edges);
            *self.app_count.entry(id).or_insert(0) += 1;
            count += 1;
        }
        if count > 0 {
            self.steps += 1;
        }
        agg.applied = count > 0;
        agg.rule_id = Some(self.grammar.rules[ri].id.clone());
        Ok(agg)
    }

    pub(crate) fn step(&mut self) -> Result<StepResult, ApiError> {
        if self.grammar.config.strategy == "maximal" {
            return match self.pick_maximal_rule() {
                Some(ri) => self.apply_maximal(ri),
                None => Ok(StepResult::default()),
            };
        }

        let Some((ri, m)) = self.pick_rule_and_match() else {
            return Ok(StepResult::default());
        };

        // Probability gate: a no-op step that still returns a ruleId so the run
        // loop retries.
        let prob = self.grammar.rules[ri].probability;
        if prob < 1.0 && self.rng.next() > prob {
            return Ok(StepResult { rule_id: Some(self.grammar.rules[ri].id.clone()), ..Default::default() });
        }

        let out = apply_rule_at(
            &mut self.host,
            &self.grammar.rules[ri],
            &m,
            Some(&mut self.rng),
            &mut self.counter,
            &mut self.id_seq,
        )?;
        let id = self.grammar.rules[ri].id.clone();
        *self.app_count.entry(id.clone()).or_insert(0) += 1;
        self.steps += 1;
        Ok(StepResult {
            applied: true,
            rule_id: Some(id),
            created_nodes: out.created_nodes,
            created_edges: out.created_edges,
            deleted_nodes: out.deleted_nodes,
            deleted_edges: out.deleted_edges,
        })
    }

    /// Run until no rule applies or a bound is hit. `max_steps`: None → use
    /// config.maxSteps; <=0 → an uncapped run with a hard safety bound.
    pub(crate) fn run(&mut self, max_steps: Option<i64>) -> Result<i64, ApiError> {
        let requested = max_steps.unwrap_or(self.grammar.config.max_steps);
        let cap: i64 = if requested > 0 { requested } else { 500_000 };
        let mut applied = 0i64;
        let mut empty_streak = 0;
        for _ in 0..cap {
            let r = self.step()?;
            if !r.applied {
                // probability skips return a ruleId; allow bounded retries.
                if r.rule_id.is_some() && empty_streak < 64 {
                    empty_streak += 1;
                    continue;
                }
                break;
            }
            empty_streak = 0;
            applied += 1;
        }
        Ok(applied)
    }
}

pub(crate) fn engine_from_json(grammar_json: &str, start_json: Option<&str>) -> Result<Engine, ApiError> {
    let grammar: Grammar = parse_validated(grammar_json, schema_check::grammar(), "grammar")?;
    let start = match start_json {
        Some(s) => Some(parse_validated::<Graph>(s, schema_check::graph(), "start graph")?),
        None => None,
    };
    Ok(Engine::new(grammar, start))
}

/// Build an engine from a grammar, run it, and return `(final_graph_json, applied)`.
/// Convenience for tests / one-shot use; the FFI uses the stateful handle.
pub fn run_grammar(
    grammar_json: &str,
    start_json: Option<&str>,
    max_steps: Option<i64>,
) -> Result<(String, i64), ApiError> {
    let mut engine = engine_from_json(grammar_json, start_json)?;
    let applied = engine.run(max_steps)?;
    let graph = serde_json::to_string(&engine.host.to_graph()).expect("graph serializes");
    Ok((graph, applied))
}
